### PR TITLE
Separate heaps

### PR DESCRIPTION
**Pull request by erikcorry**
This is almost all just separating and duplicating code.

Probably both heaps can be simplified a lot after this.

The page size is still shared between the program and object heaps.
A next step could be to make the program heap detection range
based on non-embedded so that the page sizes are not tied.

### CODE REVIEW

File: src/compiler/CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +41,7 @@ target_link_libraries(
   toit.compile
   toit_core
   toit_compiler
+  toit_core
```
Comment (kasperl) [inline src/compiler/CMakeLists.txt:44]: This is already in line 42?

File: src/primitive_core.cc [hunk 2]
```diff
@@ -2225,13 +2225,11 @@ PRIMITIVE(literal_index) {
   ARGS(Object, o);
   auto null_object = process->program()->null_object();
   if (!o->is_heap_object()) return null_object;
-  auto heap_object = HeapObject::cast(o);
-  if (!Heap::in_read_only_program_heap(heap_object, process->object_heap())) return null_object;
```
Comment (kasperl) [inline src/primitive_core.cc:2229]: Was it necessary to fix this or did you just decide to remove the fast case? I assume the fast (fail) case isn't used by toitp.

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/primitive_core.cc (hunk 2 above)
Comment (erikcorry) [reply src/primitive_core.cc:2229]: We have circular dependencies between .a files - not sure why we didn't have this issue before, but this makes it link.

Comment (erikcorry) [reply src/primitive_core.cc:2229]: Maybe it wasn't necessary.

**Pull request merged**